api/traces/v1/trace_rbac.go

```diff
@@ -0,0 +1,182 @@
+					err = tempopb.UnmarshalFromJSONV1(b, trace)
+					if err != nil {
+						return err
+					}
+					trace = traceRBAC(allowedNamespaces, trace)
+
+					traceResponseBody, err := tempopb.MarshalToJSONV1(trace)
+					if err != nil {
+						return err
+					}
+					response.Body = io.NopCloser(bytes.NewBuffer(traceResponseBody))
+				} else {
+					searchResponse := &tempopb.SearchResponse{}
+					err = jsonpb.UnmarshalString(string(b), searchResponse)
+					if err != nil {
+						return err
+					}
+					searchResponse = searchResponseRBAC(allowedNamespaces, searchResponse)
+
+					marshaller := jsonpb.Marshaler{}
+					responseBuffer := &bytes.Buffer{}
+					err = marshaller.Marshal(responseBuffer, searchResponse)
+					if err != nil {
+						return err
+					}
+					response.Body = io.NopCloser(responseBuffer)
+				}
+				// We could re-encode in gzip/deflate, but there is no need, so send it raw
+				response.Header["Content-Encoding"] = []string{}
+			}
+
+			return nil
+		}
+
+		return nil
+	}
+}
+
+func traceRBAC(allowedNamespaces map[string]bool, trace *tempopb.Trace) *tempopb.Trace {
+	for _, rs := range trace.ResourceSpans {
+		notAllowedNamespace := ""
+		if rs.Resource != nil && rs.Resource.Attributes != nil {
+			for _, resAttr := range rs.Resource.Attributes {
+				if resAttr.Key == "k8s.namespace.name" && !allowedNamespaces[resAttr.Value.GetStringValue()] {
+					notAllowedNamespace = resAttr.Value.GetStringValue()
+					for _, scopeSpan := range rs.ScopeSpans {
+						for _, span := range scopeSpan.Spans {
+							// clear all data
+							span.Attributes = []*commonv1.KeyValue{}
+							span.Events = []*tracev1.Span_Event{}
+						}
+						scopeSpan.Scope.Attributes = []*commonv1.KeyValue{}
+					}
+				}
+			}
+			// add namespace back if it was not allowed
+			if notAllowedNamespace != "" {
+				rs.Resource.Attributes = []*commonv1.KeyValue{}
+				rs.Resource.Attributes = append(rs.Resource.Attributes, &commonv1.KeyValue{
+					Key:   "k8s.namespace.name",
+					Value: &commonv1.AnyValue{Value: &commonv1.AnyValue_StringValue{StringValue: notAllowedNamespace}},
+				})
+			}
+		}
+	}
+	return trace
+}
+
+func searchResponseRBAC(allowedNamespaces map[string]bool, searchResponse *tempopb.SearchResponse) *tempopb.SearchResponse {
+	for _, traceSearchMetadata := range searchResponse.GetTraces() {
+		for i := range traceSearchMetadata.GetSpanSets() {
+			traceSearchMetadata.SpanSets[i] = spanSetRBAC(allowedNamespaces, traceSearchMetadata.SpanSets[i])
+		}
+		traceSearchMetadata.SpanSet = spanSetRBAC(allowedNamespaces, traceSearchMetadata.GetSpanSet())
+	}
+	return searchResponse
+}
+
+func spanSetRBAC(allowedNamespaces map[string]bool, spanSet *tempopb.SpanSet) *tempopb.SpanSet {
+	for _, span := range spanSet.GetSpans() {
+		notAllowedNamespace := ""
+		for _, attribute := range span.GetAttributes() {
+			if attribute.GetKey() == "k8s.namespace.name" && !allowedNamespaces[attribute.GetValue().GetStringValue()] {
```

Comment:
What happens when you add `| select(resource.k8s.namespace.name)` to the search query, but the trace doesn't have a `k8s.namespace.name` attribute? I think to be safe I'd filter in case this attribute is (for any reason) not present in the search response.

Reply:
yes I have added a test case for it